classes/observers/listeners/PKPUsageEventLogListener.inc.php

```diff
@@ -0,0 +1,305 @@
+<?php
+
+/**
+ * @file classes/observers/listeners/PKPUsageEventLogListener.inc.php
+ *
+ * Copyright (c) 2014-2021 Simon Fraser University
+ * Copyright (c) 2000-2021 John Willinsky
+ * Distributed under the GNU GPL v3. For full terms see the file docs/COPYING.
+ *
+ * @class PKPUsageEventLogListener
+ * @ingroup observers_listeners
+ *
+ * @brief Listener listening for the usage events.
+ */
+
+namespace PKP\observers\listeners;
+
+use APP\core\Application;
+use APP\facades\Repo;
+use APP\observers\events\UsageEvent;
+use APP\statistics\StatisticsHelper;
+use GeoIp2\Database\Reader;
+use PKP\cache\CacheManager;
+use PKP\cache\FileCache;
+use PKP\file\PrivateFileManager;
+use Sokil\IsoCodes\IsoCodesFactory;
+
+class PKPUsageEventLogListener
+{
+    public FileCache $geoDataCache;
+
+    /**
+     * Handle the event.
+     */
+    public function handle(UsageEvent $usageEvent): void
+    {
+        $usageEventArray = $this->prepareUsageEvent($usageEvent);
+        $this->logUsageEvent($usageEventArray);
+    }
+
+    /**
+     * Log the usage event
+     */
+    public function logUsageEvent(array $usageEventArray): void
+    {
+        $usageEventLogEntry = json_encode($usageEventArray) . PHP_EOL;
+
+        // Log file name (from the current day)
+        $logFileName = $this->getUsageEventLogFileName();
+
+        // Write the event to the log file
+        // Keep the locking in order not to care about the filesystems's block sizes and if the file is on a local filesystem
+        $fp = fopen($logFileName, 'a+b');
+        if (flock($fp, LOCK_EX)) {
+            fwrite($fp, $usageEventLogEntry);
+            flock($fp, LOCK_UN);
+        } else {
+            // Couldn't lock the file.
+            assert(false);
+        }
+        fclose($fp);
+    }
+
+    /**
+     * Prepare the usage event:
+     *  create new daily salt file, if necessary
+     *  get Geo data, if needed
+     *  get institution IDs, if needed
+     *  hash the IP
+     */
+    protected function prepareUsageEvent(UsageEvent $usageEvent): array
+    {
+        $usageEventArray = (array) $usageEvent;
```

Comment:
Hmmm... 
I thought the UsageEvent object can/should be kept as small as possible, i.e. in OJS it would now (after exchanging the object IDs with the actual objects -- see comment above) contain the following attributes:
`time, canonicalUrl, assocType, request, context, submission?, representation?, submissionFile?, issue?, issueGalley?, version`
All other information would then be retrieved in the listener (that could theoretically be queued), i.e. in our case the information needed for our log format: 
ip/hashedIp, userAgent, fileType, country, region, city, institutionIds, 
(and also: contextId, submissionId, representationId, issueId, issueGalleyId).

Thus, I do not believe we should have e.g. country, region, city as attributes in the UsageEvent class, because they are set first in our listener, and another listener could work differently (also expect those information to be there)...

Thus, I will leave it as it is now and once I implemented the changes for the UsageEvent to contain objects rather than only the object IDs, we could eventually see if we want to change this too...

Reply:
I think maybe we're confusing the Event/Listener with the log entry of the usage event we're recording. The Event/Listener shouldn't necessarily be converted directly to the log entry. That class represents the event that is fired in the event/listener system, not the usage event itself. There should be another class which represents the log entry.

Here's a rough illustration of what I mean:

```php
event(new ExampleEvent($user));
```

```php
use APP\usageEvent\Record;
use PKP\user\User;

class ExampleEvent
{
    // This would be a class that we attach all of the log details to,
    // like the user id, submission id, etc.
    protected Record $record;
    protected User $user;

    public function __construct(User $user)
    {
        $this->user = $user;
        $this->record = new Record($user);
    }
}
```

Then listeners can augment the log entry if they wan:

```php
class GeographyListener
{
    public function handle(ExampleEvent $event)
    {
        $event->record->city = $this->getCity($event->user->getIP());
    }
}
```

Then a separate listener can convert that object to the log:

```php
file_put_contents('...', json_encode($event->record), FILE_APPEND);
```

So the log entry lives separately from the event that is fired by `event()`.